### DISCUSSION TITLE
What a beautiful library

### DISCUSSION THREAD

**Discussion by candidia** [pbeshai/tidy#75] [General]
Seriously, I've gone down the rabbit hole of dataframe JS libraries today, and this is the most elegant one. It has a perfect surface area, not trying to parse CSVs when papaparse exists, not trying to reimplement console.table, not trying to handle things that aren't its responsibility. It starts with an object in memory and ends there. When you have such a clear separation of concerns, other things like streaming in data should just work because it's OUTSIDE this library. I still have a lot to learn but I'm feeling confident this is the right choice above others like arquero (which for some reason is omitting thousands of rows from my input), nodejs-polars (which wouldn't read my csv because it didn't end with .csv and doesn't work in the browser/serverless), danfo.js (which has separate APIs for browser and node), simple-data-analysis (whose scaffolded project errored), etc.

Only unfortunate thing I've encountered so far is the type issues due to dynamic columns. I think some other libraries like zod and arktype have gotten into the nitty gritty of typescript and solved things like this, but I haven't investigated deep enough to give a solution unfortunately.

Bravo

**Comment by pbeshai** [comment]
Thank you very much! I'd love to improve the types too they really annoy me, maybe some time this year I'll have a chance to give it a deep dive.
